### ISSUE TITLE
java.util.concurrent.RejectedExecutionException: event executor terminated

### ISSUE THREAD

**Issue by 1124826889** [lettuce-io/lettuce-core#1274]
## Bug Report

```
org.springframework.data.redis.RedisSystemException: Unknown redis exception; nested exception is java.util.concurrent.RejectedExecutionException: event executor terminated
        at org.springframework.data.redis.FallbackExceptionTranslationStrategy.getFallback(FallbackExceptionTranslationStrategy.java:53)
        at org.springframework.data.redis.FallbackExceptionTranslationStrategy.translate(FallbackExceptionTranslationStrategy.java:43)
        at org.springframework.data.redis.connection.lettuce.LettuceConnection.convertLettuceAccessException(LettuceConnection.java:268)
        at org.springframework.data.redis.connection.lettuce.LettuceStringCommands.convertLettuceAccessException(LettuceStringCommands.java:799)
        at org.springframework.data.redis.connection.lettuce.LettuceStringCommands.set(LettuceStringCommands.java:180)
        at org.springframework.data.redis.connection.DefaultedRedisConnection.set(DefaultedRedisConnection.java:281)
        at sun.reflect.GeneratedMethodAccessor53.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.data.redis.core.CloseSuppressingInvocationHandler.invoke(CloseSuppressingInvocationHandler.java:61)
        at com.sun.proxy.$Proxy103.set(Unknown Source)
        at com.idreamsky.dc.exactmarket.admin.service.RunService.lambda$run$0(RunService.java:151)
        at org.springframework.data.redis.core.RedisTemplate.lambda$executePipelined$1(RedisTemplate.java:324)
        at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:224)
        at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:184)
        at org.springframework.data.redis.core.RedisTemplate.execute(RedisTemplate.java:171)
        at org.springframework.data.redis.core.RedisTemplate.executePipelined(RedisTemplate.java:320)
        at org.springframework.data.redis.core.RedisTemplate.executePipelined(RedisTemplate.java:310)
        at com.idreamsky.dc.exactmarket.admin.service.RunService.run(RunService.java:145)
        at com.idreamsky.dc.exactmarket.admin.controller.ActvController.lambda$run$0(ActvController.java:80)
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1626)
        at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1618)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.util.concurrent.RejectedExecutionException: event executor terminated
        at io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:846)
        at io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:329)
        at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:322)
        at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:757)
        at io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:232)
        at io.netty.util.concurrent.AbstractScheduledEventExecutor.schedule(AbstractScheduledEventExecutor.java:155)
        at io.netty.util.concurrent.AbstractEventExecutorGroup.schedule(AbstractEventExecutorGroup.java:50)
        at io.netty.util.concurrent.AbstractEventExecutorGroup.schedule(AbstractEventExecutorGroup.java:32)
        at io.lettuce.core.protocol.CommandExpiryWriter.potentiallyExpire(CommandExpiryWriter.java:164)
        at io.lettuce.core.protocol.CommandExpiryWriter.write(CommandExpiryWriter.java:111)
        at io.lettuce.core.RedisChannelHandler.dispatch(RedisChannelHandler.java:187)
        at io.lettuce.core.StatefulRedisConnectionImpl.dispatch(StatefulRedisConnectionImpl.java:152)
        at io.lettuce.core.AbstractRedisAsyncCommands.dispatch(AbstractRedisAsyncCommands.java:467)
        at io.lettuce.core.AbstractRedisAsyncCommands.set(AbstractRedisAsyncCommands.java:1208)
        at org.springframework.data.redis.connection.lettuce.LettuceStringCommands.set(LettuceStringCommands.java:167)
        ... 21 common frames omitted

```
#### Current Behavior

<details>
<summary>Stack trace</summary>

```java
// your stack trace here;
```

</details>

#### Input Code

```java
                Expiration expiration = Expiration.from(cacheTime);
                AtomicInteger count = new AtomicInteger();
                Iterator<List<PlayerBagKey>> scrollIter = esService.scroll(getQueryDto(actvDto));
                while (scrollIter.hasNext()) {
                    List<PlayerBagKey> pbkList = scrollIter.next();
                    // (RunService.java:145) error here
                    redisUtil.getRedisTemplate().executePipelined((RedisCallback<Object>) connection -> {
                        for (PlayerBagKey pbk : pbkList) {
                            String key = CacheKeyUtil.getPbKey(pbk, dataVersion.toString());
                            String value = JSON.toJSONString(bagKey) + "_" + pbk.getVersion();
                            connection.set(key.getBytes(), value.getBytes(), expiration, RedisStringCommands.SetOption.UPSERT);
                            count.incrementAndGet();
                        }
                        return null;
                    });
                }
```

</details>

#### Expected behavior/code

the value can be set successfully
#### Environment
Lettuce version(s): [e.g. 5.1.7,RELEASE]
Redis version: [e.g. 5.0]
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-redis</artifactId>
            <version>2.1.6.RELEASE</version>
        </dependency>

**Comment by mp911de** [comment]
This looks like you’re issuing commands after shutting down the Redis client.

**Comment by mp911de** [comment]
Closing due to lack of activity. If you would like us to look at this issue, please provide the requested information and we will re-open the issue.

**Comment by martin-tarjanyi** [comment]
@mp911de we are seeing the same issue on Lettuce 5.1.8 after enabling this client option `.publishOnScheduler(true)`. Client is not shut down during commands.

**Comment by mp911de** [comment]
The issue happens when the `EventExecutorGroup` gets terminated. This happens typically on application shutdown.

**Comment by martin-tarjanyi** [comment]
In our case the same application instance produces this exception for the last one hour, now. It doesn't reject all requests, just a small portion of them. We try to upgrade to latest Spring Boot and Lettuce client to see if it makes the error go away.

**Comment by mp911de** [comment]
It makes sense to trace down from where the termination call originates that shuts down the executor. You can configure your own executor through `ClientResources` and customize the `shutdown` methods.

**Comment by quaff** [comment]
Our application is not shutdown, but the issue happens.
spring-data-redis:2.1.4
lettuce:5.1.3
netty:4.1.32

**Comment by martin-tarjanyi** [comment]
@mp911de we are still seeing this issue every now and then. This time, I have some new observation though. Just before the first `RejectedExecutionException` would arise we are seeing a `RedisCommandTimeoutException` (we have timeout configured on the client):

```
Thread=[lettuce-eventExecutorLoop-5-7] Logger=reactor.core.publisher.Operators Operator called default onErrorDropped
io.lettuce.core.RedisCommandTimeoutException: Command timed out after 100 millisecond(s)
	at io.lettuce.core.ExceptionFactory.createTimeoutException(ExceptionFactory.java:51) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.lettuce.core.protocol.CommandExpiryWriter.lambda$potentiallyExpire$0(CommandExpiryWriter.java:172) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_111]
```

```
Thread=[lettuce-eventExecutorLoop-5-7] Logger=io.netty.util.concurrent.SingleThreadEventExecutor Type=WARN Unexpected exception from an event executor: 
reactor.core.Exceptions$BubblingException: io.lettuce.core.RedisCommandTimeoutException: Command timed out after 100 millisecond(s)
	at reactor.core.Exceptions.bubble(Exceptions.java:173) ~[reactor-core-3.3.9.RELEASE.jar!/:3.3.9.RELEASE]
	at reactor.core.publisher.Operators.onErrorDropped(Operators.java:635) ~[reactor-core-3.3.9.RELEASE.jar!/:3.3.9.RELEASE]
	at reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:83) ~[reactor-core-3.3.9.RELEASE.jar!/:3.3.9.RELEASE]
	at io.lettuce.core.RedisPublisher$OnComplete.run(RedisPublisher.java:1094) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_111]
Caused by: io.lettuce.core.RedisCommandTimeoutException: Command timed out after 100 millisecond(s)
	at io.lettuce.core.ExceptionFactory.createTimeoutException(ExceptionFactory.java:51) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.lettuce.core.protocol.CommandExpiryWriter.lambda$potentiallyExpire$0(CommandExpiryWriter.java:172) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	... 5 more
```

Is it possible that the timeout causes the event executor to shut down?

**Comment by mp911de** [comment]
Lettuce doesn’t shut down event executors by itself. Ideally, wrap ClientResources and throw an exception on shutdown (or print the stack trace) to find out where this comes from.

**Comment by n3whous3** [comment]
@mp911de as I checked the codebase there and the callstacks, that exception Martin mentions is indeed a source of shutdown, since it is a generic exception handling inside the core of the event loop of that executor, which then proceeds with the finally section, with a shutdown. So that timeout of ours caused it, may it be another third party's way of exposing this exception or not, haven't checked it yet. So I didn't see indicators yet that our code tries to reuse this particular executor later on in the logs, but I suppose, so the issue might be that the system reuses a shutdown executor.
https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java#L992

But this has to be traced down in netty then.

**Comment by martin-tarjanyi** [comment]
Hi @mp911de ,

(Let us know if we should open a new ticket instead of abusing this one 😄  )

We've found two possible root causes why this could happen. At least one of these seems to be related to the lettuce code base.
### 1. Timeout handling not thread safe
If we're looking at the stack trace shared above which eventually triggers the executor shutdown:
```
Thread=[lettuce-eventExecutorLoop-5-7] Logger=reactor.core.publisher.Operators Operator called default onErrorDropped
io.lettuce.core.RedisCommandTimeoutException: Command timed out after 100 millisecond(s)
	at io.lettuce.core.ExceptionFactory.createTimeoutException(ExceptionFactory.java:51) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.lettuce.core.protocol.CommandExpiryWriter.lambda$potentiallyExpire$0(CommandExpiryWriter.java:172) ~[lettuce-core-5.3.3.RELEASE.jar!/:5.3.3.RELEASE]
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170) ~[netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) [netty-common-4.1.51.Final.jar!/:4.1.51.Final]
```
We can see that it is triggered by an `onErrorDropped` which suggests an ill-behaved `Publisher` which emits an `onError` event after an `onNext`/`onComplete`. As the error is triggered by a timeout, I started to look at how the command timeout is implemented in lettuce. I ended up in the `CommandExpiryWriter` class and especially this section: https://github.com/lettuce-io/lettuce-core/blob/26895dd19ebc3567b6aa3f555dd78eec85b2f46b/src/main/java/io/lettuce/core/protocol/CommandExpiryWriter.java#L168-L184
Now I'm not sure how I could verify that indeed this is the issue but I believe there is a race condition here between the timeout scheduler and the completion handler.
So if this line evaluates to true: https://github.com/lettuce-io/lettuce-core/blob/26895dd19ebc3567b6aa3f555dd78eec85b2f46b/src/main/java/io/lettuce/core/protocol/CommandExpiryWriter.java#L170
...then CPU schedules another task, the redis response arrives, this line will also evaluate to true: https://github.com/lettuce-io/lettuce-core/blob/26895dd19ebc3567b6aa3f555dd78eec85b2f46b/src/main/java/io/lettuce/core/protocol/CommandExpiryWriter.java#L180
This will cause that both the completion and the timeout are triggered which eventually causes the `onErrorDropped`.

### 2. Different error handling in case of executors.
Another outstanding question is: why the issue is only present in case of `publishOnScheduler(true)`?

Now, we are not entirely sure whether the issue lies in the netty or lettuce code base but it seems that errors are handled differently in the two executors or executor groups:
* If `publishOnScheduler(false)`, then Nio event loop group is used which has a quite complex implementation and I'm not sure how are the errors are handled in that case. They might be handled internally?
* If `publishOnScheduler(true)` then DefaultEventExecutor is used which does not seem to do any error handling and just lets the exceptions pop out which eventually kills the executor as it can be seen in the stack trace.

**Comment by ujhazib** [comment]
And as an add-on to Martin's observation, even if this is the case and it is fixed, that's Netty's problem right now that an executor group is not able to heal itself, which I see it is as a missing design in their API. Opened an issue for that, hopefully they can give some hint how this should be handled better or if they want to include this into their design: https://github.com/netty/netty/issues/10918

**Comment by mp911de** [comment]
The `isDone` check is an optimization to avoid exception creation if the command is already done. There's no race. Calling `complete*(…)` methods multiple times makes additional calls a no-op.

Note that reactive operators are a consequence of completing a command successfully/exceptionally. As mentioned above, completion is guaranteed to happen only once. Everything else would be a violation of the reactive streams spec.

Regarding your second question:
With `publishOnScheduler(false)`, all operator calls (unless you're using `flatMap`, `concatMap` or `subscribeOn` operators) happen on the EventLoop (I/O) thread and therefore synchronously.

Using `publishOnScheduler(true)` offloads reactive signals to a worker thread/queue and therefore decouples activity from the I/O thread. There's also no way how to attach/synchronize reactive signals with the I/O thread so I can imagine pretty well that shutting down the client can still lead to the situation where there's work in the worker queue while the EventLoop is already stopped.

I currently don't see how Lettuce can solve that issue. It's more an orchestration issue that requires stopping incoming requests and completing ongoing requests before you can shut down Lettuce.

**Comment by ujhazib** [comment]
> I currently don't see how Lettuce can solve that issue. It's more an orchestration issue that requires stopping incoming requests and completing ongoing requests before you can shut down Lettuce.

Problem is that this is not a full shut down, but as mentioned above, a bubbling exception cause that ONE of the netty's executor is shut down and that executor still remains in the executor group. From the callstacks the obvious path what Netty took there was that a publisher found itself being called when the done variable was already true, hence something called it twice. Martin's code part was an idea how it could happen, since  the issue on our side always started a timeout and that code part is doing a scheduled execution and registers also an oncomplete. Idea was that both of them called in a race condition and none of them finds isDone being true, so both runs down.

**Comment by martin-tarjanyi** [comment]
@mp911de 
I was able to create a repro, see: https://github.com/martin-tarjanyi/lettuce-executor-reject-repro

I'm not sure we are on the same page regarding the timeout issue. The presence of `Operators.onErrorDropped` in the stacktrace indicates a reactive stream violation (the one that you also mentioned #7) if I'm not mistaken: https://github.com/reactive-streams/reactive-streams-jvm#1-publisher-code The timeout exception seems to be coming after the onComplete signal was delivered. It is possible though that it is coming from a different place than I initially thought but it is still coming from Lettuce in some way based on the stack trace.

Also, note that we are not shutting down the client/template/executor. The bubbling exception is the one which causes the executor shutdown and the eventual constant failure.

You'll see in the repro that setting `publishOnScheduler(false)` makes the error go away. Also, we've discovered that updating Reactor 3.4.x fixes the issue as well, because they changed the default behaviour of `onErrorDropped` and it's just logging and not throwing anymore.

Still, we believe at the root of the issue there is a reactive stream violation and an error handling discrepancy.

**Comment by mp911de** [comment]
Looking at `SubscriptionCommand` and `CommandWrapper`, they do not check whether the command was completed so both implementations are error-prone for the spec violation. Can you file a bug report referencing this ticket so we can fix the issue ASAP?
